dev/sg/migration.go

```diff
@@ -437,3 +446,209 @@ func readFilenamesNamesInDirectory(dir string) ([]string, error) {
 
 	return names, nil
 }
+
+type migration struct {
+	ID   int
+	Name string
+
+	UpName   string
+	DownName string
+}
+
+type migrationConflict struct {
+	ID     int
+	Trunk  migration
+	Branch migration
+}
+
+// trunk is the branch/revisikon that we are branched from. It's the good one. Usually it would be main
+func fixupMigrations(database Database, trunk, branch string) error {
+	out.Write("")
+	trunkMigrations, err := getMigrationsForRevision(database, trunk)
+	if err != nil {
+		return err
+	}
+
+	branchMigrations, err := getMigrationsForRevision(database, branch)
+	if err != nil {
+		return err
+	}
+
+	block := out.Block(output.Linef(output.EmojiLightbulb, output.StyleItalic, "Checking for conflicting migrations..."))
+	defer block.Close()
+
+	conflicts, err := findConflictingMigrations(trunkMigrations, branchMigrations)
+	if err != nil {
+		return err
+	}
+
+	if len(conflicts) == 0 {
+		block.WriteLine(output.Linef(output.EmojiSuccess, output.StyleReset, "... No conflicting migrations"))
+		return nil
+	}
+
+	if err := resolveConflictingMigrations(database, conflicts, trunkMigrations, block); err != nil {
+		return err
+	}
+
+	return err
+}
+
+func findConflictingMigrations(trunkMigrations, branchMigrations map[int]migration) ([]migrationConflict, error) {
+	conflicts := []migrationConflict{}
+
+	for migrationID, trunkMigration := range trunkMigrations {
+		branchMigration, ok := branchMigrations[migrationID]
+		if !ok {
+			return nil, errors.Newf(
+				"It appears you haven't rebased off of your trunk. Rebase first and then run again. You are missing: %s",
+				trunkMigration.Name,
+			)
+		}
+
+		if trunkMigration.Name != branchMigration.Name {
+			conflicts = append(conflicts, migrationConflict{
+				ID:     migrationID,
+				Trunk:  trunkMigration,
+				Branch: branchMigration,
+			})
+		}
+	}
+
+	return conflicts, nil
+}
+
+func getMigrationsForRevision(database Database, revision string) (map[int]migration, error) {
+	baseDir, err := migrationDirectoryForDatabase(database)
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := runGitCmd("ls-tree", "--name-only", "-r", revision, baseDir)
+	if err != nil {
+		return nil, err
+	}
+
+	upMigrations := make(map[int]string)
+	downMigrations := make(map[int]string)
+
+	files := strings.Split(output, "\n")
+	for _, file := range files {
+		if file == "" {
+			continue
+		}
+
+		migrationID, ok := parseMigrationIndex(file)
+		if !ok {
+			return nil, errors.New("Bad migration file format")
+		}
+
+		if strings.HasSuffix(file, ".down.sql") {
+			downMigrations[migrationID] = file
+		} else if strings.HasSuffix(file, ".up.sql") {
+			upMigrations[migrationID] = file
+		} else {
+			return nil, errors.New("Somehow have a file that doesn't end with up or down")
```

Comment:
```suggestion
			return nil, errors.New("Invalid filename: you have a file that doesn't end with '.up.sql' or '.down.sql'")
```
Also would be nice if you included the problematic filename in the error message.

Reply:
I'd skip this, we might have readmes and stuff in these dirs.

Reply:
Oh, yeah, good point.